ambry-store/src/main/java/com.github.ambry.store/IndexSegment.java

```diff
@@ -456,13 +420,16 @@ public int getNumberOfItems() {
    *  key n / value n - the key and value entries contained in this index segment
    *  crc             - the crc of the index segment content
    *
-   * @param safeEndPoint
+   * @param safeEndPoint the end point until which it is safe to flush
    * @throws IOException
    * @throws StoreException
    */
-  public void writeIndexToFile(long safeEndPoint) throws IOException, StoreException {
-    if (prevSegmentEndOffset != safeEndPoint) {
-      if (safeEndPoint > getEndOffset()) {
+  void writeIndexToFile(Offset safeEndPoint) throws IOException, StoreException {
+    if (safeEndPoint.compareTo(startOffset) < 0) {
+      return;
+    }
+    if (!safeEndPoint.equals(prevSegmentEndOffset)) {
```

Comment:
what really is `prevSegmentEndOffset`? I am pretty sure it is badly and confusingly named. For a while I thought this referred to an offset in the previous segment. 
Seems like it is used to figure out whether there have been any entries since the last time this file was persisted. This should probably be called `prevSafeEndPoint` or `persistedTillOffset` or something. This change I think should go in this PR itself.